pandas/_libs/src/klib/khash_python.h

```diff
@@ -330,6 +333,129 @@ Py_hash_t PANDAS_INLINE tupleobject_hash(PyTupleObject* key) {
 }
 
 
+// TODO: if nanos is most common/important, might be most performant to
+// make that canonical and others cast to that?
+Py_hash_t PANDAS_INLINE hash_datetime_value_and_reso(npy_datetime value, NPY_DATETIMEUNIT unit, npy_datetimestruct* dts) {
+    // If we cannot cast to pydatetime, then the question is if there are
+    // other-resolution datetime64 objects that we might be equal to whose
+    // hashes we need to match. We let year-reso objects return value, and make
+    // higher-resolution cases responsible for checking of they match.
+    if (unit == NPY_FR_Y) {
+        return value;
+    }
+    else if (unit == NPY_FR_M) {
+        if ((value % 12) == 0) {
+            return hash_datetime_value_and_reso(value / 12, NPY_FR_Y, dts);
```

Comment:
So is the idea here that we want things to compare to equal across different frequencies?

Reply:
correct.  if a == b then hash(a) should equal hash(b); https://github.com/numpy/numpy/issues/3836

Reply:
Hmm this seems a bit tricky. I see your point about the equality but am also seeing this comment about hashing across different units:

https://github.com/numpy/numpy/issues/3836#issuecomment-25577274

So not sure where I stand on this. Was it decided upstream that this is in fact the behavior we wanted? My initial thought was that unless things were in the same frequency I wouldn't want them hashed to the same value

Reply:
> but am also seeing this comment about hashing across different units:

that's handled here.  if you can convert to a lower reso losslessly (i.e. you are == to a lower-reso dt64 object) then do so and give the hash of that.  otherwise you are in the clear to return your own i8 value

> My initial thought was that unless things were in the same frequency I wouldn't want them hashed to the same value

Having x == y without hash(x) == hash(y) causes a cascade of problems.  I encountered this when trying to make pd.unique preserve non-nano dt64s and tracked the ensuing test failures back here.

Reply:
> and then multiply always rather than divide, start from ns and trust that overflows are OK

That's kind of like what I did for `Timestamp.__hash__` (if we go with this, I'll need to change `Timestamp.__hash__` to match this).  Tried casting to a higher-resolution and checked for overflows.  Couldn't figure out how to do the same thing outside of cython.